### DISCUSSION TITLE
Could 3 nodes, 1 driver per node ensure HA?

### DISCUSSION THREAD

**Discussion by Clara12062** [minio/minio#18572] [Q&A] [answered]
Hi!
&emsp; Due to environmental limitations, I deployed a 3-node MinIO setup with a single drive per node in a Kubernetes cluster. I referred to the deployment documentation: https://min.io/docs/minio/kubernetes/upstream/operations/concepts/architecture.html#minio-architecture. 
&emsp; The documentation mentions a minimum of 4 minio hosts. I understand that 4 drivers is the minimum required for high availability. And I'm unsure my understanding is correct. Because in the 3-node single-driver environment, I found that there is also "HA". I don't know how minio HA is implemented anymore. I'm unclear about how MinIO achieves high availability in this scenario.

### Testing steps:
1. Upload an object to the MinIO cluster.
2. Power off one of the nodes.
3. Use `mc cp ALIAS/bucket/object ./` to download the object.
4. Check the MD5 of the downloaded object.

### Observation:
The object is successfully downloaded. And md5 value is correct.

### Expected Result:
If high availability requires a minimum of 4 drives, then in a 3-drive setup with one drive offline, the object should be corrupted and not downloadable.

**Comment by ravindk89** [comment, marked as answer]
When we discuss recommended minimums in our docs, we're talking about what we feel comfortable as a low watermark for a minimum viable deployment.

Before continuing, take a look at [this page of our docs](https://min.io/docs/minio/linux/reference/minio-server/settings/storage-class.html#envvar.MINIO_STORAGE_CLASS_STANDARD). Admittedly it's a little buried but it goes over the default EC (parity) values for a given erasure set size.

Also please review our [Erasure Code Core Concepts](https://min.io/docs/minio/kubernetes/upstream/operations/concepts/erasure-coding.html).

The Erasure Set Size + Erasure Parity are what impact how many drives or nodes can go down while maintaining quorum.

With a 4x4 cluster, you'd have 1 erasure set of 16 drives by default and EC:4. That is, you can lose up to 4 drives while the cluster is online and still handle reads/writes.

That's equivalent to *one* node at most going down. With EC:8 you can lose up to *two* nodes and at least maintain reads. Lets take a note of that.

Now, with a 4 x 1 deployment topo, you have 4 drives in your erasure set - so you are defaulting to EC:2 at a baseline. Same basic level of HA - you can lose at most two nodes and still maintain your data.

Take us to a 3 x 1 now - EC:1. You can lose no more than one drive - and therefore one node - before you risk losing data.

A 3x4 gets you to 12 drives and a default EC of 4, but you still can't lose more than two nodes - your second node going down for any reason would immediately result in quorum loss. 


In comparison, 4 node distribution has *higher availability* than 3 node given the same approximate number of drives.  And even that, losing two nodes still puts you in the danger zone. 

That's why 4 is our "minimum". We have an [open PR](https://github.com/minio/docs/pull/1080) that will actually update our guidance to suggest 8 nodes as a baseline in production, primarily because the more nodes you have (within reason) the less likely it is that even a handful of nodes going down can meaningfully impact the cluster in the short term.

I'll also note - the Operator is a management layer in Kubernetes for MinIO.

MinIO can technically start in distributed mode with as few as 2 nodes, but we would *never* recommend that.

>  The total number of MinIO server pods to deploy in the Tenant. The Operator enforces a minimum of four server pods per tenant.

That is an Operator-based limiter intended to guide users based on best practices.

**Reply by Clara12062** [reply]
Thank you very much for your detailed reply! :smile:

I get the idea. 3 x 1 deployments are riskier, right? 
I have another question. Do 3x1 deployment and 3x4 deployment have the same risk of data loss?

**Reply by harshavardhana** [reply]
Yes

**Reply by harshavardhana** [reply]
HA is a function of number of nodes you have, since nodes going down is also as common as drives going bad.

**Reply by Clara12062** [reply]
Thanks for your reply! 
i kinda get it.
